tests/switch.t

```diff
@@ -0,0 +1,144 @@
+local test = require 'test'
+
+function dotest(fn, a, b, c, d)
+  test.eq(fn(1), a)
+  test.eq(fn(2), b)
+  test.eq(fn(3), c)
+  test.eq(fn(4), d)
+end
+
+terra foo(a: int): int
+  switch a do
+    case 1 then
+      return 2
+    case 2 then
+      return 5
+    end
```

Comment:
Just to check, is this end optional? I can see why we allow it (for consistency with spliced quotes), but it seems nicer to be able to elide it:

```
switch a do
  case 1 then
    return 2
  case 2 then
    return 5
  else
    return 3
  end
end
```

Reply:
The end is currently mandatory, but I do not remember why - @aiverson might remember. We were running into parsing issues with how terra handles blocks which greatly limited our ability to make parts of the syntax optional.

Reply:
The syntax you gave in the example is incorrect. You have one too many  `end`s.
`else` doesn't match a case, because that would make the syntax not enforce the 0 or 1 default branches constraint and make linearizing the case statements in the presence of metaprogramming be much harder to write. The else clause is part of the switch statement, not the case clauses. `switch expr do caselist else block end`. Case statements terminate at an `end` consuming the keyword, or at a `case` or `else` not consuming the keyword.

Therefore, you should be able to write
 ```terra
switch a do
  case 1 then
    return 2
  case 2 then
    return 5
else
  return 3
end
```
without an issue, though I don't think the tests currently enforce that. Putting an `end` after the final case like
 ```terra
switch a do
  case 1 then
    return 2
  case 2 then
    return 5
  end
else
  return 3
end
```
is closer to the intended syntax. The `end` there is needed in order to be able to put splices containing additional cases at the end of a list of cases in a switch statement, like
```terra
switch a do
  case 1 then
    return 2
  end
  [generate_case_list(stuff)]
else
  return 3
end
```
which might be useful for tuning instruction locality.

Reply:
Ok, I think I understand now. Will reply below, this seems like an important point that I don't want to get lost in the discussion.